### PR TITLE
SOLR-15217: Use shardsWhitelist in ReplicationHandler.

### PR DESCRIPTION
**Pull request by bruno-roustant**
I didn't update the doc as I'm not sure where/what to update. Should I?

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/handler/IndexFetcher.java [hunk 1]
```diff
@@ -271,7 +273,33 @@ public IndexFetcher(@SuppressWarnings({"rawtypes"})final NamedList initArgs, fin
     String httpBasicAuthPassword = (String) initArgs.get(HttpClientUtil.PROP_BASIC_AUTH_PASS);
     myHttpClient = createHttpClient(solrCore, httpBasicAuthUser, httpBasicAuthPassword, useExternalCompression);
   }
-  
+
+  private void setLeaderUrl(String leaderUrl) {
+    if (leaderUrl != null && solrCore.getCoreContainer().getShardHandlerFactory() instanceof HttpShardHandlerFactory) {
```
Comment (atris) [inline solr/core/src/java/org/apache/solr/handler/IndexFetcher.java:278]: NIT: Lets get the ZKController once and reuse?

**Review by atris — CHANGES_REQUESTED** [review]

File: solr/core/src/java/org/apache/solr/handler/IndexFetcher.java [hunk 2]
```diff
@@ -271,7 +274,36 @@ public IndexFetcher(@SuppressWarnings({"rawtypes"})final NamedList initArgs, fin
...
+            throw new SolrException(ErrorCode.BAD_REQUEST,
+                "Invalid URL syntax in '" + LEADER_URL + "' with value '" + leaderUrl + "'", e);
+          } else {
+            log.warn("The '{}' parameter value '{}' is not in the '{}'",
+                LEADER_URL, leaderUrl, HttpShardHandlerFactory.INIT_SHARDS_WHITELIST);
+            throw new SolrException(SolrException.ErrorCode.FORBIDDEN,
+                "The '" + LEADER_URL + "' parameter value '" + leaderUrl
+                    + "' is not in the '" + HttpShardHandlerFactory.INIT_SHARDS_WHITELIST + "'."
+                    + HttpShardHandlerFactory.SET_SOLR_DISABLE_SHARDS_WHITELIST_CLUE);
+          }
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/handler/IndexFetcher.java:300]: I'm not a fan of log-and-throw.

File: solr/core/src/java/org/apache/solr/handler/IndexFetcher.java [hunk 3]
```diff
@@ -271,7 +274,36 @@ public IndexFetcher(@SuppressWarnings({"rawtypes"})final NamedList initArgs, fin
...
+          if (e.code() == ErrorCode.BAD_REQUEST.code) {
+            throw new SolrException(ErrorCode.BAD_REQUEST,
+                "Invalid URL syntax in '" + LEADER_URL + "' with value '" + leaderUrl + "'", e);
+          } else {
+            log.warn("The '{}' parameter value '{}' is not in the '{}'",
+                LEADER_URL, leaderUrl, HttpShardHandlerFactory.INIT_SHARDS_WHITELIST);
+            throw new SolrException(SolrException.ErrorCode.FORBIDDEN,
+                "The '" + LEADER_URL + "' parameter value '" + leaderUrl
+                    + "' is not in the '" + HttpShardHandlerFactory.INIT_SHARDS_WHITELIST + "'."
+                    + HttpShardHandlerFactory.SET_SOLR_DISABLE_SHARDS_WHITELIST_CLUE);
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/handler/IndexFetcher.java:299]: Is suggesting how to disable the shards whitelist the most prudent strategy here? Maybe a link to the docs is safer?

I feel like this is going to hurt an operator on a page somewhere.

File: solr/core/src/test/org/apache/solr/handler/TestReplicationHandlerDiskOverFlow.java [hunk 4]
```diff
@@ -67,6 +67,7 @@
 
   @Before
   public void setUp() throws Exception {
+    systemSetPropertySolrDisableShardsWhitelist("true");
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/handler/TestReplicationHandlerDiskOverFlow.java:70]: same question here, re: deprecation

**Review by madrob — COMMENTED** [review]
I have a lot of other thoughts about this too, but I think most of those are better addressed as part of the bigger change in 9.0 and this change should be minimally scoped. A few questions/comments left in the review, but otherwise I think this is pretty safe for back port.

File: solr/core/src/test/org/apache/solr/handler/TestReplicationHandler.java [hunk 5]
```diff
@@ -302,6 +305,26 @@ public void doTestHandlerPathUnchanged() throws Exception {
     assertEquals("/replication", ReplicationHandler.PATH);
   }
 
+  @Test
+  public void testShardsWhitelist() throws Exception {
+    // Run another test with shards whitelist enabled and whitelist is empty.
+    // Expect an exception because the leader URL is not allowed.
+    systemClearPropertySolrDisableShardsWhitelist();
```
[the remark this answers is not in the document]
Comment (bruno-roustant) [reply solr/core/src/test/org/apache/solr/handler/TestReplicationHandler.java:312]: For backward compatibility I have added systemSetPropertySolrDisableShardsWhitelist("true") in setUp() and systemClearPropertySolrDisableShardsWhitelist() in tearDown() as this is already done in other tests (otherwise most test methods fail).
So here the idea is to have one method that does not disable shard whitelisting. Another approach could be to move systemSetPropertySolrDisableShardsWhitelist() to each test method, but this doesn't improve readibility nor scalability if we increase the number of test cases. I would prefer to keep this call here as we are effectively in a test class that requires backqard compatibility about whitelisting.

File: solr/core/src/java/org/apache/solr/handler/IndexFetcher.java (hunk 2 above)
Comment (bruno-roustant) [reply solr/core/src/java/org/apache/solr/handler/IndexFetcher.java:300]: I copied this code from the existing HttpShardHandlerFactory.WhitelistHostChecker.checkWhiteLlist().
Yes that seems redundant. I'll remove in both places.

File: solr/core/src/java/org/apache/solr/handler/IndexFetcher.java (hunk 3 above)
Comment (bruno-roustant) [reply solr/core/src/java/org/apache/solr/handler/IndexFetcher.java:299]: This suggestion exists in 3 other places: HttpShardHandler, HttpShardHandlerFactory, TermsComponent.
I propose we keep it for this minimal patch for branch 8x and we can add a link to the doc in the change for main 9x?

File: solr/core/src/test/org/apache/solr/handler/TestReplicationHandlerDiskOverFlow.java (hunk 4 above)
Comment (bruno-roustant) [reply solr/core/src/test/org/apache/solr/handler/TestReplicationHandlerDiskOverFlow.java:70]: This is also for backward compatibility for an existing test. The same pattern is used for 10 other existing tests.
Maybe I can try to replace that by a dynamic solr config in the 9x patch, but should we do that here?

**Review by dsmiley — APPROVED** [review]
I looked through the change and also the other reviewer comments.  I think mostly the particulars are there mirroring existing code/tests.

File: solr/core/src/test/org/apache/solr/handler/TestReplicationHandlerDiskOverFlow.java (hunk 4 above)
Comment (bruno-roustant) [reply solr/core/src/test/org/apache/solr/handler/TestReplicationHandlerDiskOverFlow.java]: Finally I could remove this particular use of the deprecated method in favour of settings the whitelist (which is not possible in TestReplicationHandler because it creates JettySolrRunner later so we don't know the ports ahead of time).

File: solr/core/src/java/org/apache/solr/handler/IndexFetcher.java [hunk 6]
```diff
@@ -271,7 +274,34 @@ public IndexFetcher(@SuppressWarnings({"rawtypes"})final NamedList initArgs, fin
...
+
+  private void setLeaderUrl(String leaderUrl) {
+    if (leaderUrl != null) {
+      ShardHandlerFactory shardHandlerFactory = solrCore.getCoreContainer().getShardHandlerFactory();
+      if (shardHandlerFactory instanceof HttpShardHandlerFactory) {
+        ZkController zkController = solrCore.getCoreContainer().getZkController();
+        ClusterState clusterState = zkController == null ? null : zkController.getClusterState();
+        try {
+          ((HttpShardHandlerFactory) shardHandlerFactory).getWhitelistHostChecker()
+              .checkWhitelist(clusterState, null, Collections.singletonList(leaderUrl));
```
Comment (cholpas) [inline solr/core/src/java/org/apache/solr/handler/IndexFetcher.java:286]: When using Solr with only leader/follower replication and not with SolrCloud, this whitelist check is failing - unless `-Dsolr.disable.shardsWhitelist=true` is passed. In my point of view it does not make sense to have to pass this flag when SolrCloud is not being used. What do you think?

Comment (janhoy) [reply solr/core/src/java/org/apache/solr/handler/IndexFetcher.java:286]: This is by design and on purpose, please see documentation of the feature here https://solr.apache.org/guide/8_8/distributed-requests.html#configuring-the-shardhandlerfactory

In leader/follower installs, it is expected that you configure the `shardsWhitelist` setting in solr.xml with a list of all your shards, to avoid user queries with shards parameter being able to connect to any host on the network.

Comment (cholpas) [reply solr/core/src/java/org/apache/solr/handler/IndexFetcher.java:286]: I'm not using distributed search, just old style master/slave or in new terms leader/follower replication. There is no configuration about shards, SolrCloud or whatsoever, but this whitelist check throws the following exception: `Invalid URL syntax in "shards" parameter: null`

I don't believe that this is the desired behaviour when configuring only a `solr.ReplicationHandler` in solrconfig.xml. In the documentation https://solr.apache.org/guide/8_8/index-replication.html there is no word mentioned about shards.

Comment (janhoy) [reply solr/core/src/java/org/apache/solr/handler/IndexFetcher.java:286]: Ok, this is for validation of `leaderUrl` in replication handler. In SOLR-15340 the setting is renamed to `allowUrls` which better reflects the generic intent for it, such as validating any user/config provided URL in the system.

As this sounds like a bug, please file a NEW Jira issue with a clear description on how to reproduce the issue, including the Exception stacktrace you see, so we can get a bug fix in the next bugfix release.

Comment (cholpas) [reply solr/core/src/java/org/apache/solr/handler/IndexFetcher.java:286]: Hmm, yes... I tried now tried twice to create a jira ticket and spent quite some time formatting it nicely. Both times (with different IPs) when submitting the ticket I got `Your request has been blocked. If you feel this is in error, please let us know at: abuse@infra.apache.org`. Srsly, wtf... I already sent an email there, but didn't get a response.
